File: AGC033/A.py
from collections import deque


def solve():
    H, W = map(int, input().split())
    grid = [['' for _ in range(W)] for _ in range(H)]
    que = deque()
    for i in range(H):
        grid[i] = list(input())
        for j in range(W):
            if grid[i][j] == '#':
                que.append([i, j])
    
    xy = [[0, -1], [1, 0], [0, 1], [-1, 0]]
    ans = 0
    while que:
        n = len(que)
        for _ in range(n):
            y, x = que.popleft()
            for i in range(4):
                xi = x - xy[i][0]
                yi = y - xy[i][1]
                if 0 <= xi <= W-1 and 0 <= yi <= H-1:
                    if grid[yi][xi] == '.':
                        grid[yi][xi] = '#'
                        que.append([yi, xi])
        ans += 1
    print(ans-1)


if __name__ == '__main__':
    solve()

File: AGC033/A.nim
import strutils, sequtils, queues

proc solve() =
    var H, W: int
    (H, W) = stdin.readLine.strip.split.map(parseInt)
    var grid = mapIt(0..<H, stdin.readLine)

    type P = object
        y, x: int
    var que = initQueue[P]()
    for i in 0..<H:
        for j in 0..<W:
            if grid[i][j] == '#':
                que.enqueue(P(y: i, x: j))
    
    let dyx = [[-1, 0], [0, 1], [1, 0], [0, -1]]
    var ans = 0
    while que.len != 0:
        var n = que.len
        for _ in 0..<n:
            let cur = que.dequeue
            for i in 0..3:
                var
                    ny = cur.y + dyx[i][0]
                    nx = cur.x + dyx[i][1]
                if 0 <= ny and ny < H and 0 <= nx and nx < W:
                    if grid[ny][nx] == '.':
                        grid[ny][nx] = '#'
                        que.enqueue(P(y: ny, x: nx))
        ans += 1
    echo ans-1

solve()